File: ashcore/benchmarks/bench_arena.py
"""
Benchmark: Python memory allocation — 1M x 64B bytearrays.
Equivalent to bench_arena.mojo / bench_arena.c.
"""
import gc
import time

N = 1_000_000

# warm up
for _ in range(1000):
    _ = bytearray(64)

gc.disable()

t0 = time.perf_counter_ns()
chunks = [bytearray(64) for _ in range(N)]
t1 = time.perf_counter_ns()
del chunks
t2 = time.perf_counter_ns()

gc.enable()

print(f"alloc_total_ns={t1 - t0}")
print(f"free_ns={t2 - t1}")
print(f"per_op_ns={(t1 - t0) // N}")
print(f"n={N}")

File: ashcore/benchmarks/bench_arena.c
/*
 * Benchmark: malloc/free — 1M x 64B allocs, then bulk free.
 * Equivalent to bench_arena.mojo for comparison.
 */
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static long ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000000000L + ts.tv_nsec;
}

int main(void) {
    int N = 1000000;
    void **ptrs = malloc(N * sizeof(void *));
    if (!ptrs) { fputs("OOM\n", stderr); return 1; }

    /* warm up */
    for (int i = 0; i < 1000; i++) { void *p = malloc(64); free(p); }

    long t0 = ns();
    for (int i = 0; i < N; i++)
        ptrs[i] = malloc(64);
    long t1 = ns();
    for (int i = 0; i < N; i++)
        free(ptrs[i]);
    free(ptrs);
    long t2 = ns();

    printf("alloc_total_ns=%ld\n", t1 - t0);
    printf("free_ns=%ld\n",        t2 - t1);
    printf("per_op_ns=%ld\n",      (t1 - t0) / N);
    printf("n=%d\n",               N);
    return 0;
}
